apps/emqx_rule_engine/src/emqx_rule_events.erl

```diff
@@ -1060,7 +1060,7 @@ printable_maps(Headers) ->
             (K, V, AccIn) ->
                 AccIn#{K => V}
         end,
-        #{},
+        #{'User-Property' => #{}, 'User-Property-Pairs' => []},
```

Comment:
```suggestion
        #{'User-Property' => #{}},
```

Reply:
If I understand it correctly, you'd like to ensure `User-Property` in the output of this function ?

Reply:
But I'm not quite following why you'd always need the `User-Property` field.

Reply:
Actually, MQTT 5 allows duplicated property keys,
`User-Property-Pairs` was invented to keep the original (potentially duplicated keys).
Maybe we should use this filed instead `User-Property` to restore the Properties for downstream MQTT message.

Reply:
Then we need to decide how to merge `pub_props.'User-Property'` and `pub_props.'User-Property-Pairs'` after the SQL query runs. It is more convenient to read/modify `pub_props.'User-Property'`, since it is just a map rather than an array (I use it in my EMQX setup, to prevent looped messages between a pair of egress&ingress bridges between two brokers). So either:
- Leave only `'User-Property-Pairs'`; not convenient to work w/ arrays; lack of array functions
- Leave only `'User-Property'`, lacks duplicated properties feature
- Keep both, but use `'User-Property'` for the packet construction; `'User-Property-Pairs'` would be read-only thing (current behavior)
- Rewrite the `'User-Property'` structure to `#{string() -> list(string())}`; again, working with arrays, but allows duplicates
- Some other struct, convenient to work with from Rules' SQL

Reply:
> 



> But I'm not quite following why you'd always need the `User-Property` field.

If there's no `'User-Property'` field in `pub_props`, then an expression like
```sql
select
  *,
  map_put('key', 'val', pub_props.'User-Property') as pub_props.'User-Property'
from "mytopic/#"
```
would fail, wouldn't it?
We can also check for empty map in `'User-Property'`, after SQL execution; if it is so, then omit it completely in a packet serialization.

Reply:
convinced that using 'User-Property' is the most straightforward approach.